feat: add information about readonly types in generated TypeScript (#692)
vaadin/docs: articles/fusion/application/accessing-backend.asciidoc

== Example TypeScript Module Contents

For example, the generated TypeScript module for the Java endpoint defined in
<<accessing-backend#how-to-create-vaadin-endpoint,CounterEndpoint.java>> would look as follows:

.`CounterEndpoint.ts`
[source,typescript]
----
include::{root}/frontend/generated/CounterEndpoint.ts[]
----

== Entities

> Endpoint methods can return or receive as a parameter any data structure (entity).

Reviewer: ```suggestion
An Endpoint method can return or receive a parameter as an entity (non-primitive type).
```

> In this case, the TypeScript generator creates a TypeScript interface for it.

Reviewer: ```suggestion
In this case, the generator also creates a TypeScript interface for the entity.
```

`City.java`
[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/fusion/accessingbackend/City.java[]
----

> `CountryEndpoint.java`

Reviewer: Could remove this, it looks redundant on the page.

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/fusion/accessingbackend/CountryEndpoint.java[]
----

The TypeScript output is the following:

> `City.ts`

Reviewer: Could remove this, it looks redundant on the page.
![Screenshot 2021-08-13 at 11 01 13](https://user-images.githubusercontent.com/30408303/129324646-632f10cf-9a58-4588-8e3d-04df9bb387d7.png)

Author: Oh, nice catch. Why they appeared this way, I wonder.

[source,typescript]
----
include::{root}/frontend/generated/com/vaadin/demo/fusion/accessingbackend/City.ts[]
----

> `CountryEndpoint.ts`

Reviewer: Could remove this, it looks redundant on the page.

Reviewer: Or maybe we could return the whole CountryEndpoint source code? It's kinda repetitive since we showed the CountEndpoint.
Might be better to keep the Entity section about Entity interface generation, not to repeat the Endpoint generation.

Author: I wanted to show how the endpoint with an entity is generated. I believe it is better to show the code here because only the code for an entity might look unclear.

 * @module CounterEndpoint
 */

import client from './connect-client.default';

/**
 * A method that adds one to the argument.
 *
 * @param number
 */
function _addOne(number: number): Promise<number> {
  return client.call('CounterEndpoint', 'addOne', {number});
}

export {
  _addOne as addOne,
}
----

== Entities

Endpoint methods can return or receive as a parameter any data structure (entity).
In this case, the TypeScript generator creates a TypeScript interface for it.
Let's take a look at the example.

> `City.java`

Reviewer: Since these are newly added source code, it's better to put the code into separate files and reference to them via `include::`,  e.g.
```
.`SecurityConfig.java`
[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/fusion/security/authentication/SecurityConfigDemo.java[tags=public-resources]
----
```
in https://github.com/vaadin/docs/edit/latest/articles/fusion/security/spring-login.asciidoc

Author: Looks like code moved to a separate file is compiled as an executable code while my examples are only visual. So I guess it is better to leave inlined.
